app/grants/clr.py

```diff
@@ -181,27 +182,60 @@ def get_totals_by_pair(contrib_dict):
 
         returns:
             pair totals between current round
-                {user_id (str): {user_id (str): pair_total (float)}}
-
+                {
+                    user_id (str): {
+                        user_id (str): pair_total (float)
+                    }
+                }
+            curr_agg_sqrts
+                {
+                    grantId (int): {
+                        user_id (float): sqrt(aggregated_amount) (float)
+                    }
+                }
+            pairs_tot
+                int: the number of cycles it took to construct the pairs
     '''
+    pairs_tot = 0
     pair_totals = {}
-
+    curr_agg_sqrts = {}
     # start pairwise match
-    for _, contribz in contrib_dict.items():
+    for proj, contribz in curr_agg.items():
```

Comment:
fwiw, i have a *hunch* that much of the aggregation (and sqrt + pairwise) code could be rewritten in SQL (stays in the DB server) and be much much much more efficient than pulling it all back from the database server + assembling it in memory on the task server + then writing it back to the db.

its just a hunch tho.

Reply:
I had the same suspicion and wrote this version: https://github.com/gitcoinco/web/blob/53707dc28231563f24fe6e3737d5c6a6a48693bb/app/grants/clr2.py

(POC here: https://dbfiddle.uk/?rdbms=postgres_9.6&fiddle=ba238edb1599780f776099ed98bcb2e4)

I couldn't prove it was more efficient than making the pairs + clr in python though :( My findings are over here: https://github.com/gitcoinco/web/pull/9348 - there could be some optimisations we could make that I've missed though?